client/my-sites/plans-features-main/plan-notices.tsx

```diff
@@ -0,0 +1,136 @@
+	);
+
+	const isHigherPlanAvailable = function () {
+		const visiblePlansWithoutEnterprise = visiblePlanNames.filter(
+			( planName ) => planName !== PLAN_ENTERPRISE_GRID_WPCOM
+		);
+		const highestPlanName = visiblePlansWithoutEnterprise.pop();
+		return highestPlanName !== currentSitePlanSlug;
+	};
+
+	const isUpgradeEligibleSite =
+		isSiteOnPaidPlan && isNotJetpackSiteOrIsAtomicSite && isHigherPlanAvailable();
+
+	return {
+		creditsValue,
+		isPlanUpgradeCreditEligible: isUpgradeEligibleSite && creditsValue > 0,
+	};
+}
+
+export default function PlanNotices( {
+	siteId,
+	isInSignup,
+	visiblePlanNames = [],
+	discountInformation: { withDiscount, discountEndDate },
+}: {
+	visiblePlanNames: string[];
+	isInSignup: boolean;
+	siteId: number;
+	discountInformation: {
+		withDiscount: string;
+		discountEndDate: Date;
+	};
+} ) {
+	const translate = useTranslate();
+	const [ isNoticeDismissed, setIsNoticeDismissed ] = useState( false );
+	const currencyCode = useSelector( ( state ) => getCurrentUserCurrencyCode( state ) );
+	const canUserPurchasePlan = useSelector(
+		( state ) =>
+			! isCurrentPlanPaid( state, siteId ) || isCurrentUserCurrentPlanOwner( state, siteId )
+	);
+	const activeDiscount = getDiscountByName( withDiscount, discountEndDate );
+	const { creditsValue, isPlanUpgradeCreditEligible } = usePlanUpgradeCreditsDisplay(
+		siteId,
+		visiblePlanNames
+	);
+
+	if ( isNoticeDismissed ) {
+		return null;
+	} else if ( ! canUserPurchasePlan ) {
+		return (
+			<Notice
+				className="plan-features__notice"
+				showDismiss={ true }
+				onDismissClick={ () => setIsNoticeDismissed( true ) }
+				status="is-info"
+			>
+				{ translate(
+					'This plan was purchased by a different WordPress.com account. To manage this plan, log in to that account or contact the account owner.'
+				) }
+			</Notice>
+		);
+	} else if ( activeDiscount ) {
+		return (
+			<Notice
+				className="plan-features__notice-credits"
+				showDismiss={ true }
+				onDismissClick={ () => setIsNoticeDismissed( true ) }
+				icon="info-outline"
+				status="is-success"
+			>
+				{ activeDiscount?.plansPageNoticeTextTitle && (
+					<strong>
+						{ activeDiscount?.plansPageNoticeTextTitle }
+						<br />
+					</strong>
+				) }
+				{ activeDiscount.plansPageNoticeText }
+			</Notice>
+		);
+	} else if ( isPlanUpgradeCreditEligible && ! isInSignup ) {
+		return (
+			<Notice
+				className="plan-features__notice-credits"
+				showDismiss={ true }
+				onDismissClick={ () => setIsNoticeDismissed( true ) }
+				icon="info-outline"
+				status="is-success"
+			>
+				{ translate(
+					'You have {{b}}%(amountInCurrency)s{{/b}} of pro-rated credits available from your current plan. ' +
+						'Apply those credits towards an upgrade before they expire!',
+					{
+						args: {
+							amountInCurrency: formatCurrency( creditsValue, currencyCode ?? '' ),
+						},
+						components: {
+							b: <strong />,
+						},
+					}
+				) }
+			</Notice>
+		);
+	}
```

Comment:
I think they are in order. At least that is what I gathered from the previous implementation.

https://github.com/Automattic/wp-calypso/blob/af1a71803031f7d3e139201e3c3e50850625c81c/client/my-sites/plan-features-2023-grid/index.tsx#L903-L916

We are trying to populate 1 notice in a single location. So there is bound to be an order of precedence if not those conditions do not make sense. This is the order that I have explicitly signalled with the if else chain :)

Reply:
I think having this as an explicit if else chain will be best. We can possibly move the contents of each if else chain into descriptive child components though. But the cyclomatic complexity will still remain which is I think the main source of complexity here.

I will add some tests here as well so that we are guarded from possibly regressions.